lib/mayaUsd/undo/UsdUndoStateDelegate.cpp

```diff
@@ -0,0 +1,573 @@
+//
+// Copyright 2020 Autodesk
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+#include "UsdUndoStateDelegate.h"
+
+#include "UsdUndoBlock.h"
+#include "UsdUndoManager.h"
+
+#include <mayaUsd/base/debugCodes.h>
+
+PXR_NAMESPACE_USING_DIRECTIVE
+
+namespace MAYAUSD_NS_DEF {
+
+UsdUndoStateDelegate::UsdUndoStateDelegate()
+    : _dirty(false)
+    , _setMessageAlreadyShowed(false)
+{
+    // TfDebug::Enable(USDMAYA_UNDOSTATEDELEGATE);
+}
+
+UsdUndoStateDelegate::~UsdUndoStateDelegate() { }
+
+UsdUndoStateDelegateRefPtr UsdUndoStateDelegate::New()
+{
+    return TfCreateRefPtr(new UsdUndoStateDelegate());
+}
+
+bool UsdUndoStateDelegate::_IsDirty() { return _dirty; }
+
+void UsdUndoStateDelegate::_MarkCurrentStateAsClean() { _dirty = false; }
+
+void UsdUndoStateDelegate::_MarkCurrentStateAsDirty() { _dirty = true; }
+
+void UsdUndoStateDelegate::_OnSetLayer(const SdfLayerHandle& layer)
+{
+    if (layer) {
+        TF_DEBUG(USDMAYA_UNDOSTATEDELEGATE)
+            .Msg("Setting Layer '%s' \n", layer->GetDisplayName().c_str());
+        _layer = layer;
+    } else {
+        _layer = nullptr;
+    }
+}
+
+void UsdUndoStateDelegate::invertSetField(
+    const SdfPath& path,
+    const TfToken& fieldName,
+    const VtValue& inverse)
+{
+    _setMessageAlreadyShowed = true;
+
+    TF_DEBUG(USDMAYA_UNDOSTATEDELEGATE)
+        .Msg("Inverting set Field '%s' for Spec '%s'\n", fieldName.GetText(), path.GetText());
+
+    SetField(path, fieldName, inverse);
+
+    _setMessageAlreadyShowed = false;
```

Comment:
I was hoping you won't pick this pattern as a solution to the problem. Consider improving because it's very possible that this reporting will break with time when someone adds more code in here.

Reply:
Yup, I will improve it later. The reason didn't do it was because I was very lazy last night and this meant changing more code. Should be a trivial change.